worlds/blasphemous/Rules.py

```diff
@@ -1,1479 +1,3333 @@
+        elif boss == "Mothers":
+            return has_boss_strength("melquiades") and \
+                (self.has("D04Z02S22[W]", player) or \
+                    self.has("D04Z02S22[E]", player))
+        elif boss == "Canvases":
+            return has_boss_strength("exposito") and \
+                (self.has("D05Z02S14[W]", player) or \
+                    self.has("D05Z02S14[E]", player))
+        elif boss == "Prison":
+            return has_boss_strength("quirce") and \
+                (self.has("D09Z01S03[W]", player) or \
+                    self.has("D09Z01S03[N]", player))
+        elif boss == "Rooftops":
+            return has_boss_strength("crisanta") and \
+                (self.has("D06Z01S25[W]", player) or \
+                    self.has("D06Z01S25[E]", player))
+        elif boss == "Ossuary":
+            return has_boss_strength("isidora") and \
+                self.has("D01BZ08S01[W]", player)
+        elif boss == "Mourning":
+            return has_boss_strength("sierpes") and \
+                self.has("D20Z02S08[E]", player)
+        elif boss == "Graveyard":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D02Z03S23[E]", player) and \
+                        self.has("D02Z02S14[W]", player) and \
+                            self.has("Wall Climb Ability", player)
+        elif boss == "Jondo":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    (self.has("D20Z01S05[W]", player) or \
+                        self.has("D20Z01S05[E]", player)) and \
+                            (self.has("D03Z01S03[W]", player) or \
+                                self.has("D03Z01S03[SW]", player))
+        elif boss == "Patio":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D06Z01S18[E]", player) and \
+                        (self.has("D04Z01S04[W]", player) or \
+                            self.has("D04Z01S04[E]", player)) or \
+                                self.has("D04Z01S04[Cherubs]", player)
+        elif boss == "Wall":
+            return has_boss_strength("amanecida") and \
+                self.has("D01BZ07S01[Santos]", player) and \
+                    self.has("D09BZ01S01[Cell24]", player) and \
+                        (self.has("D09Z01S01[W]", player) or \
+                            self.has("D09Z01S01[E]", player))
+        elif boss == "Hall":
+            return has_boss_strength("laudes") and \
+                (self.has("D08Z03S03[W]", player) or \
+                    self.has("D08Z03S03[E]", player))
+        elif boss == "Perpetua":
+            return has_boss_strength("perpetua")
+        elif boss == "Legionary":
+            return has_boss_strength("legionary")
```

Comment:
This could maybe be a dictionary too, if you're clever about it. Something like a dict of tuples, where tuple[0] is the boss and tuple[1] is a list of (event?) item names. As in:

`boss_rules = Dict[Tuple[str,List[str]
...
return has_boss_strength(boss_rules[boss][0]) and all(self.has(item, player) for item in boss_rules[boss][1])`

Although you might have to be careful about "capturing" variables in this case? Not sure.

Reply:
I see what you're saying but I don't think this would work because several of the bosses' requirements have multiple acceptable conditions rather than just all of the above items.